### ISSUE TITLE
[Question] I do not underestanc the GPU and memory usage of SB3

### ISSUE THREAD

**Issue by EloyAnguiano** [DLR-RM/stable-baselines3#1630]
### ❓ Question

I think I do not underestand the memory usage of SB3. I have a Dict observation space of some huge matrixes, so my observation space is 17MB approx:
```
(Pdb) [sys.getsizeof(v) for k, v in obs.items()]
[2039312, 2968, 12235248, 105800, 2968, 2968, 2968, 2039312, 116, 2039312, 2968, 2968, 2968]
(Pdb) sum([sys.getsizeof(v) for k, v in obs.items()])/1024/1024
17.623783111572266
```
I am making a Vectorized environment with the make_vec_env function of size 2 and the hyperparameters of my PPO agents are n_steps = 6 and my batch_size is 16. If I underestood well, my rollout buffer will be `n_steps x n_envs = 12` so the rollout_buffer will be `17 x 12 = 204 MB`. I assume that the `batch_size` of `16` will get the minimum so it is equivalent of having a batch size of `12`.

The problem here is that when I'm using a GPU device (80GB A100) it stabilizes at 70GB of usage at the beginning and a little bit later it stops for the lack of space at the device. How is this even possible?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
there is an important information missing which is your network architecture.
The rollout buffer store things in the RAM not on the GPU.
And most GPU memory is taken by weights and gradients.

Might be a duplicate of https://github.com/DLR-RM/stable-baselines3/issues/863

**Comment by EloyAnguiano** [comment]
Printing my model size with this:
```
def print_model_size(model):
    param_size = 0
    for param in model.parameters():
        param_size += param.nelement() * param.element_size()
    buffer_size = 0
    for buffer in model.buffers():
        buffer_size += buffer.nelement() * buffer.element_size()

    size_all_mb = (param_size + buffer_size) / 1024**2
    print(f'model size: {size_all_mb:.3f}MB')  # noqa: T201
```
And calling it like:
`print_model_size(agent.policy)`

Returns:
`model size: 8.369MB`

Is there any part of the agent that could be bigger? I am using my custom `FeatureExtractor` class but I assume its included at policy argument

**Comment by EloyAnguiano** [comment]
Also, whenever yoou run a batch od data on GPU ypu have to transfer those data to the cuda device, so the data are in GPU at some point, isn't it?

**Comment by EloyAnguiano** [comment]
@araffin I think that GPU usage could be a bit more optimal. First of all, debugging the PPO class (the train method) I found that  the GPU usage just before anything else, if I keep every hyperparameter fixed (n_steps, batch_size, etc...) but I change the number of environments at the vectorized environment, the GPU usage differs from one another:

1 environment:  1815MiB
16 environments: 8551MiB

I do not understand this as the `self.rollout_buffer.size()` is still the `n_steps` as before (32 in my case), so I do not know where does this come from. Indeed the only specifications that should affect the memory usage should be the policy size itself, the batch_size (this is key, the rollout_buffer should be at RAM and whenever we want to train with a batch, you retrieve those data to GPU) and the gradients of the model for the backpropagation. 

Does this make any sense? Am I missing something?

**Comment by araffin** [comment]
This should answer your question:

https://github.com/DLR-RM/stable-baselines3/blob/aab545901fe331814f822060d677f22191cba419/stable_baselines3/common/buffers.py#L391-L398

**Comment by EloyAnguiano** [comment]
Yes, it does. Thanks a lot. Also, this opens up my ather question and it is that I think that the rollout buffer should not be on GPU and that the GPU usage should be controlled by the batch size at each epoch. Thus, you could collect a giant rollout_buffer but train on a small but fast GPU by choosing a correct batch_size. Isn't it?

**Comment by araffin** [comment]
https://github.com/DLR-RM/stable-baselines3/pull/1720#issuecomment-1776700957

**Comment by EloyAnguiano** [comment]
Sorry, I do not understand. If rollout buffer is always on cpu, why at https://github.com/DLR-RM/stable-baselines3/issues/1630#issuecomment-1775419847 the number of used environments improves the GPU usage?

**Comment by EloyAnguiano** [comment]
Indeed, if I debug PPO training at a GPU, I get this:
```
(Pdb) self.rollout_buffer.device
device(type='cuda', index=2)
```

**Comment by araffin** [comment]
> the number of used environments improves the GPU usage?

are you using subprocesses? if so, that might be due to the way python multiprocessing work.

> This should mean that the data of the rollout_buffer are alocated at GPU

if you look at the code (and you should), the device is only used here:
https://github.com/DLR-RM/stable-baselines3/blob/aab545901fe331814f822060d677f22191cba419/stable_baselines3/common/buffers.py#L127-L139

when sampling the data there:
https://github.com/DLR-RM/stable-baselines3/blob/aab545901fe331814f822060d677f22191cba419/stable_baselines3/common/buffers.py#L520

**Comment by EloyAnguiano** [comment]
I am creating the environment like this:
```
gym_env = make_vec_env(make_env,
                               env_kwargs=env_kwargs,
                               n_envs=args.n_envs,
                               vec_env_cls=SubprocVecEnv)
```
So I assume it uses some kind of multiprocessing, yes. What does this has to do with GPU usage?

**Comment by EloyAnguiano** [comment]
Hi again @araffin . I am still unable to figure out how, it the transition of data from RolloutBuffer is done at each sampling, how can the GPU usage be so big just when the code goes into train method, as this should not have any data on GPU, only the model.

**Comment by EloyAnguiano** [comment]
I just saw that on DictRolloutBuffer, the returned samples batch have to do with the number of environments as well:
https://github.com/DLR-RM/stable-baselines3/blob/aab545901fe331814f822060d677f22191cba419/stable_baselines3/common/buffers.py#L823
When Someone puts a batch_size for its training I think that someone is trying to fit exactly those samples on that step, not `batch_size * n_envs`, isn't it?
